### ISSUE TITLE
Single Resource Fetching

### ISSUE THREAD

**Issue by tada5hi** [nuxt/nuxt#20293]
### Environment

------------------------------
- Operating System: `Windows_NT`
- Node Version:     `v16.16.0`
- Nuxt Version:     `3.3.3`
- Nitro Version:    `2.3.2`
- Package Manager:  `npm@8.11.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

```vue
<script lang="ts">
export default defineNuxtComponent({
    async setup() {
         const route = useRoute();

        const entity : Ref<User> = ref(null) as any;

        try {
            entity.value = await useAPI()
                .user
                .getOne(route.params.id as string, { fields: ['+email'] });
        } catch (e) {
            await navigateTo({ path: '/admin/users' });
            return;
        }

        console.log(entity.value.id); // illegal access if no return is previously defined
   }
});
</script>
<template>
         <div>{{ entity.id}}</div> <!-- no typing here due union return type of setup -->
</template>
```

### Describe the bug

It should be possible to do single resource fetching in a nuxt componente, when requesting the API it should also be possible to catch situations where the request was not successful.
It is possible to catch the unexpected error case with a catch, but the following code will be executed anyway, if no return statement follows.
With return statement subsequent code execution can be avoided, but the typing in the template block is no longer present since the setup routine now returns the expected context type as well as void/undefined (depending on what is returned in an error situation). 

The example in the docs refers only to successful single resource fetches.
https://nuxt.com/docs/getting-started/data-fetching#example-4

**Comment by danielroe** [comment]
I think we already do this: https://nuxt.com/docs/getting-started/error-handling#example-1.

What would you add or change about the docs?

**Comment by tada5hi** [comment]
@danielroe No, I think the docs are sufficiently described for the case of throwing just an error. But a redirect (navigateTo) is not intended or can a redirect be combined with an thrown error ( without showing an error screen for a short amount of time or preventing the redirect to happen at all) ?

**Comment by danielroe** [comment]
Have you tried it? It's documented here: https://nuxt.com/docs/api/utils/navigate-to#examples.

You can also specify a redirect within `createError` by specifying a status code. (You will also need to call `setResponseHeader` with the Location header _before_ throwing the error.)

**Comment by tada5hi** [comment]
> Have you tried it? It's documented here: https://nuxt.com/docs/api/utils/navigate-to#examples.
> 
> You can also specify a redirect within `createError` by specifying a status code. (You will also need to call `setResponseHeader` with the Location header _before_ throwing the error.)

I have tried your suggestion and indeed no error window is displayed and the redirect is correct. However, the console now displays an "unhandled error" error.

```vue
<script setup>

import { createError, navigateTo } from '#app';

const api = useAPI();
try {
    await api.realm.getOne('foo');
} catch (e) {
    await navigateTo({
        path: '/',
    });

    throw createError('Crashed');
}

console.log('foo');

</script>
```

![image](https://user-images.githubusercontent.com/13162758/232315025-b53205b3-7136-42e3-b64f-70296799e71d.png)

**Comment by nam0107** [comment]
@tada5hi I have the same problem, did you resolve it?

**Comment by tada5hi** [comment]
> @tada5hi I have the same problem, did you resolve it? 

Nope, no solution yet :/

**Comment by Hebilicious** [comment]
There's an uncaught error because you're throwing out of catch block. 
How about [this](https://play.vuejs.org/#eNp9Uk1PwzAM/StWTkWaymG3sU0CtAMcAAHHXKrU7TKypMrH6FT1v+Ok67pJQA5tYz8/P7+6Y/dNkx8CsgVbOmFl40EVul5x5h1na67lvjHWQwclVlLjo6G7Ru1nYLGCHipr9sAZUXDGNbYJTdgiqPS+rMk6roFO4Y5agEMfmuwGTsF0hNHOA0GlP8Iqtlh2slyA81bqul9nOih1c8f1VOPt8YqCzlCfHwoVkFg6iBScbdGSSOgncA+i8GILGV7JIMR0iZKMwlyZOrskzmUZhYwwS9NYTb1O4ocu9Owjank7mLtmMzKWKCtZ5ztnNPmeGnMmyCWp0L42XlJLzhajJM4Kpcz3c4p5G3A2xsUWxdcv8Z1rY4yzN4sO7YHGPud8YWv0Q3rz8YItfZ+Te1MGReh/ku9IdoSocYA9BF2S7AtcUvuU9oZ+2qfbtB61G4eKQiOyT/i0OXFB/hp9kjvP56mOLGX9Dwwt32A=) ?

**Comment by tada5hi** [comment]
@Hebilicious i think that you didn't get the issue 🙈. 
The problem is that you don't want to render or execute any further code if the entity does not exist. If you call navigateTo with a return value in the catch block, you don't get any typing for template variables

**Comment by Hebilicious** [comment]
> @Hebilicious i think that you didn't get the issue 🙈. The problem is that you don't want to render or execute any further code if the entity does not exist. If you call navigateTo with a return value in the catch block, you don't get any typing for template variables

It's much easier to help with a reproduction 🙈 You could try throwing something that is not an error, or refactor something that is not a middleware. Interestingly, I was recently talking with Daniel about allowing `throw navigateTo("/")` semantics.

**Comment by tada5hi** [comment]
@Hebilicious no i haven't tried to throw a non error object/value. The semantic you consulted with Daniel about sounds interesting. What is the status of it ?

**Comment by cernymatej** [comment]
@danielroe I think this can be closed as a duplicate of https://github.com/nuxt/nuxt/issues/23698

**Comment by danielroe** [comment]
nuxtathon closed @cernymatej
